mlc_llm/quantization/smoothquant_utils.py

```diff
@@ -0,0 +1,322 @@
+    #  * 3rd dimension - scale(multiplier) dimension.
+    a_stat = None
+    w_stat = None
+
+    target = tvm.target.Target.current(allow_none=False)
+    for data in tqdm(dataset, desc="Calibration"):
+        # Create KV-cache
+        kv_caches = kvc()
+        num_tokens = data.shape[1]
+        seq_len_shape = tvm.runtime.ShapeTuple([num_tokens])
+
+        # Run Encoder and update statistics for activations/weights
+        (logits, kv_caches), outputs = prefill(data, seq_len_shape, kv_caches, *params)
+        a_stat = _accumulate_act_outlier_stat(a_stat, outputs)
+        w_stat = _accumulate_weight_outlier_stat(w_stat, outputs)
+
+        # Run Decoder and update statistics for activations/weights
+        for _ in range(config["decoder_invoke_num"]):
+            # TODO: support softmax with temperature.
+            logits_max = np.argmax(logits.numpy(), axis=-1).astype("int32")
+            if logits_max[0] in config["stop_tokens"]:
+                break
+            next_token = tvm.nd.array(logits_max, device=tvm.device(target.kind.default_keys[0]))
+            num_tokens += logits_max.shape[1]
+            seq_len_shape = tvm.runtime.ShapeTuple([num_tokens])
+            (logits, kv_caches), outputs = decode(next_token, seq_len_shape, kv_caches, *params)
+            a_stat = _accumulate_act_outlier_stat(a_stat, outputs)
+            w_stat = _accumulate_weight_outlier_stat(w_stat, outputs)
+
+    a_stat = [np.max(s, axis=0) for s in a_stat]
+    w_stat = [np.max(s, axis=0) for s in w_stat]
+    # Use the same statistics for "prefill"/"decode"
+    stat = dict.fromkeys(funcs)
+    stat["prefill"] = (a_stat, w_stat)
+    stat["decode"] = (a_stat, w_stat)
+    for fname in funcs:
+        scale_params = _calculate_quant_scale_params(fname, stat, config, tvm.cpu(0))
+        mod = relax.transform.BindParams(fname, scale_params)(mod)
+
+    mod = mlc_llm.transform.SmoothQuantOpConverter("quantize")(mod)
+    mod = mlc_llm.transform.SmoothQuantLegalizer(config["adtype"], config["wdtype"])(mod)
+    mod = relax.transform.DeadCodeElimination(funcs)(mod)
+    return mod
+
+
+def smoothquant(args, mod, model_names):
+    target = args.target
+    smq_device = tvm.device(target.kind.default_keys[0])
+    assert args.build_model_only is False, "build_model_only in True is not supported in SMQ"
+    params = load_params(args.artifact_path, device=smq_device)
+
+    dataset, stop_tokens = _get_dummy_dataset(args.artifact_path, device=smq_device)
+    smq_config: Dict[str, Any] = {}
+    smq_config["decoder_invoke_num"] = 5
+    smq_config["alpha"] = 0.5
+    smq_config["stop_tokens"] = stop_tokens
+    smq_config["adtype"] = "int8"
+    smq_config["wdtype"] = "int8"
+    with target:
+        print("[SmoothQuant] Run smoothing...")
+        mod = _smooth(mod, params, model_names, dataset, smq_config)
+        print("[SmoothQuant] Run calibration and quantization...")
+        mod = _calibrate(mod, params, model_names, dataset, smq_config)
```

Comment:
Do we need to collect stats twice? Can't we use the first stats + smoothing scale to calculate smoothed max?

Reply:
**Short answer**: for current quantization scheme (_per-tensor_) we can call it once.

In the previous comment I mentioned that I dump preprocessed tensors (`R.absmax`, for memory footprint optimization). For per-channel or per-token quantization  we need call it twice. Or we need to dump the whole tensor, but I do not think it is a good idea.

Reply:
To demonstrate this here is the same example from above:
We have `R.linear` op and input data (activations, 2D tensor):
[[10, 1, 1],
[1, 20, 1],
[1, 1, 30]]
In the current implementation I use `R.absmax` and dump 1D tensor: [10, 20, 30].
But for _groupwise, per-token_ (or _per-channel_) quantization scheme looks like this 1D tensor is not enough to calculate all scales. As I understand we need the whole 2D tensor.

So, why do I like this op so much... :)
From my real life experiments with LLM on **A10g** GPU I met cases when I do not fit into **22 GB** of memory. As a result program crashed.
So, I decided to focus on memory optimization instead of calibration time optimization.
But as I sad, for per-tensor Q scheme this 1D vector is enough to calculate scale. And we can collect stats once.

**I would like to say the following:**
If this issue is critical (collect stats twice) I can invest part of my time and try to fix it (single run + memory optimization).
What do you think?

Reply:
> If this issue is critical (collect stats twice) I can invest part of my time and try to fix it (single run + memory optimization)

Given that you are only supporting per-tensor for now and that memory optimization is only done, I assume that single run + memory optimization doesn't require any non-trivial engineering. So I'd say why not.

That said, I still don't understand why we ever need to collect twice, even for per channel / token. We can keep track of "running max" as we send new data to the model. After all data are consumed, we are left with per-channel or token "max vector", obtained by one pass.